src/crepe/system/InputSystem.cpp

```diff
@@ -0,0 +1,183 @@
+				event_mgr.queue_event<KeyPressEvent>(KeyPressEvent{
+					.repeat = event.key_repeat,
+					.key = event.key,
+				});
+				break;
+			case SDLContext::EventType::KEYUP:
+				event_mgr.queue_event<KeyReleaseEvent>(KeyReleaseEvent{
+					.key = event.key,
+				});
+				break;
+			case SDLContext::EventType::MOUSEDOWN:
+				if (!mouse_in_viewport) {
+					break;
+				}
+				event_mgr.queue_event<MousePressEvent>(MousePressEvent{
+					.mouse_x = world_mouse_x,
+					.mouse_y = world_mouse_y,
+					.button = event.mouse_button,
+				});
+				last_mouse_down_position = {world_mouse_x, world_mouse_y};
+				last_mouse_button = event.mouse_button;
+				break;
+			case SDLContext::EventType::MOUSEUP: {
+				if (!mouse_in_viewport) {
+					break;
+				}
+				event_mgr.queue_event<MouseReleaseEvent>(MouseReleaseEvent{
+					.mouse_x = world_mouse_x,
+					.mouse_y = world_mouse_y,
+					.button = event.mouse_button,
+				});
+				//check if its a click by checking the last button down
+				int delta_x = world_mouse_x - last_mouse_down_position.first;
+				int delta_y = world_mouse_y - last_mouse_down_position.second;
+
+				if (last_mouse_button == event.mouse_button
+					&& std::abs(delta_x) <= click_tolerance
+					&& std::abs(delta_y) <= click_tolerance) {
+					event_mgr.queue_event<MouseClickEvent>(MouseClickEvent{
+						.mouse_x = world_mouse_x,
+						.mouse_y = world_mouse_y,
+						.button = event.mouse_button,
+					});
+
+					handle_click(event.mouse_button, world_mouse_x, world_mouse_y);
+				}
+			} break;
+			case SDLContext::EventType::MOUSEMOVE:
+				if (!mouse_in_viewport) {
+					break;
+				}
+				event_mgr.queue_event<MouseMoveEvent>(MouseMoveEvent{
+					.mouse_x = world_mouse_x,
+					.mouse_y = world_mouse_y,
+					.rel_x = event.rel_mouse_move.first,
+					.rel_y = event.rel_mouse_move.second,
+				});
+				handle_move(event, world_mouse_x, world_mouse_y);
+				break;
+			case SDLContext::EventType::MOUSEWHEEL:
+				event_mgr.queue_event<MouseScrollEvent>(MouseScrollEvent{
+					.scroll_x = event.wheel_delta,
+					.scroll_y = 0,
+					.direction = event.wheel_delta,
+				});
+				break;
+			case SDLContext::EventType::SHUTDOWN:
+				event_mgr.queue_event<ShutDownEvent>(ShutDownEvent{});
+				break;
+			default:
+				break;
+		}
+	}
+}
+void InputSystem::handle_move(const SDLContext::EventData & event_data,
+							  const int & world_mouse_x, const int & world_mouse_y) {
+	ComponentManager & mgr = this->component_manager;
+
+	RefVector<Button> buttons = mgr.get_components_by_type<Button>();
+
+	for (Button & button : buttons) {
+		RefVector<Transform> transform_vec
+			= mgr.get_components_by_id<Transform>(button.game_object_id);
+		Transform & transform(transform_vec.front().get());
+
+		bool was_hovering = button.hover;
+		if (button.active
+			&& is_mouse_inside_button(world_mouse_x, world_mouse_y, button, transform)) {
+			button.hover = true;
+			if (!was_hovering && button.on_enter) {
+				button.on_enter();
```

Comment:
Standaard zijn deze functies gelijk aan `nullptr`, moet daar niet op gecontroleerd worden of maakt de constructor van `std::function` automatisch een lege functie aan in dat geval?

Reply:
dit wordt wel gechecked door de button.on_enter (dalijk button.on_mouse_enter)

Reply:
Kun je dan alsjeblieft expliciet op `!= nullptr` checken zodat het iets duidelijker is? Je kunt ook aan het begin van `InputSystem::handle_button_press`
```cpp
if (button.on_enter == nullptr) return;
```
zetten en de rest van de if statements iets eenvoudiger maken